src/shaders.rs

```diff
@@ -109,234 +111,194 @@ pub fn full_shaders(device: &Device, engine: &mut WgpuEngine) -> Result<FullShad
-            BindType::BufReadOnly,
-            BindType::Buffer,
-        ],
-    )?;
-    let pathtag_scan_large = engine.add_shader(
-        device,
-        "pathtag_scan",
-        preprocess::preprocess(shader!("pathtag_scan"), &full_config, &imports).into(),
-        &[
-            BindType::Uniform,
-            BindType::BufReadOnly,
-            BindType::BufReadOnly,
-            BindType::Buffer,
-        ],
-    )?;
-    let bbox_clear = engine.add_shader(
-        device,
-        "bbox_clear",
-        preprocess::preprocess(shader!("bbox_clear"), &empty, &imports).into(),
-        &[BindType::Uniform, BindType::Buffer],
-    )?;
-    let flatten = engine.add_shader(
-        device,
-        "flatten",
-        preprocess::preprocess(shader!("flatten"), &full_config, &imports).into(),
-        &[
-            BindType::Uniform,
-            BindType::BufReadOnly,
-            BindType::BufReadOnly,
-            BindType::Buffer,
-            BindType::Buffer,
-            BindType::Buffer,
-        ],
-    )?;
-    let draw_reduce = engine.add_shader(
-        device,
-        "draw_reduce",
-        preprocess::preprocess(shader!("draw_reduce"), &empty, &imports).into(),
-        &[BindType::Uniform, BindType::BufReadOnly, BindType::Buffer],
-    )?;
-    let draw_leaf = engine.add_shader(
-        device,
-        "draw_leaf",
-        preprocess::preprocess(shader!("draw_leaf"), &empty, &imports).into(),
-        &[
-            BindType::Uniform,
-            BindType::BufReadOnly,
-            BindType::BufReadOnly,
-            BindType::BufReadOnly,
-            BindType::Buffer,
-            BindType::Buffer,
-            BindType::Buffer,
-        ],
-    )?;
-    let clip_reduce = engine.add_shader(
-        device,
-        "clip_reduce",
-        preprocess::preprocess(shader!("clip_reduce"), &empty, &imports).into(),
-        &[
-            BindType::BufReadOnly,
-            BindType::BufReadOnly,
-            BindType::Buffer,
-            BindType::Buffer,
-        ],
-    )?;
-    let clip_leaf = engine.add_shader(
-        device,
-        "clip_leaf",
-        preprocess::preprocess(shader!("clip_leaf"), &empty, &imports).into(),
-        &[
-            BindType::Uniform,
-            BindType::BufReadOnly,
-            BindType::BufReadOnly,
-            BindType::BufReadOnly,
-            BindType::BufReadOnly,
-            BindType::Buffer,
-            BindType::Buffer,
-        ],
-    )?;
-    let binning = engine.add_shader(
-        device,
-        "binning",
-        preprocess::preprocess(shader!("binning"), &empty, &imports).into(),
-        &[
-            BindType::Uniform,
-            BindType::BufReadOnly,
-            BindType::BufReadOnly,
-            BindType::BufReadOnly,
-            BindType::Buffer,
-            BindType::Buffer,
-            BindType::Buffer,
-            BindType::Buffer,
+
+    let mut force_gpu = false;
+
+    let force_gpu_from: Option<&str> = None;
```

Comment:
I think IWBN to decide this dynamically at render time. I think I'd like to move towards a mode where the presence of the CPU shaders and the decision on their execution are decoupled from each other.

A legitimate production case for this is small scenes: when rasterizing a handful of draw objects, we may want to process them quickly on the CPU to avoid the GPU upload and dispatch overhead and only run the coarse+fine raster stages on the GPU. We may want to make the CPU vs GPU decision dynamically at render time based on the size of the workload. In that world we may want to compile all the pipelines up front even though we'll sometimes use the CPU versions.

Ideally I'd like `FullShaders` to hold all available pipelines and have an option we give to `engine` that tells it which stages should run on the CPU.

Another thing I'd like you to be aware of is that we want to rewrite `src/shaders.rs` to be in terms of `crates/shaders` which automatically preprocesses the WGSL and the pipeline layouts. The CPU shaders will need to integrate with that somehow and they may move to the shaders crate (though we haven't put a lot of thought into this yet).

What you have is fine for now but be aware that this is all in flux and will change eventually.

Reply:
I would be nicer if we could run a set of shaders on the CPU, but we'll need readback hooked up on wasm first. This is a bit of a hack until then.

Reply:
I think the most common case will involve CPU stages that are strictly ordered *before* any GPU stages (rather than a mix of them), which shouldn't require read-back. I don't know if/when we'd want to run them in combination except maybe for testing.